4_LinkedListPalindrome.cpp

```diff
@@ -0,0 +1,218 @@
+Explained as comments in the code
+
+6. I - IMPLEMENT
+*/
+#include <bits/stdc++.h>
+using namespace std;
+
+struct Node 
+{
+    char data;
+    struct Node* next;
+};
+
+// Helper function to insert a new node in the beginning of the linked list
+void insert(struct Node** headRef, int data)
+{
+    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
+    node->data = data;
+    node->next = *headRef;
+    *headRef = node;
+}
+
+// Iterative function to reverse the nodes of linked list
+void reverse(struct Node** headRef)
+{
+    struct Node* reversed = NULL;
+    struct Node* current = *headRef;
+
+	// Iterate through the list and move/insert each node to the front of the result list.
+    while (current != NULL)
+	{
+        struct Node* next = current->next;
+        // move the current node to the reversed list
+        current->next = reversed;
+        reversed = current;
+        // process next node
+        current = next;
+    }
+
+    // fix head pointer
+    *headRef = reversed;
+}
+
+// Recursive function to check if two linked lists are equal or not
+bool compare(struct Node* a, struct Node* b)
+{
+    // check if both list is empty
+    if (a == NULL && b == NULL)
+        return true;
+
+    return a && b && (a->data == b->data) && compare(a->next, b->next);
+}
+
+// Function to split the linked list into two equal parts and return pointer to the second half
+struct Node* findMid(struct Node* head, bool &odd)
+{
+    struct Node* prev = NULL;
+    struct Node *slow = head, *fast = head;
+
+    // find middle pointer
+    while (fast && fast->next)
+	{
+        prev = slow;
+        slow = slow->next;
+        fast = fast->next->next;
+    }
+
+    // for odd nodes, fast still points to last node
+    if (fast)
+	    odd = true;
+
+    // make next of prev node null
+    prev->next = NULL;
+
+    // return mid node: the slow pointer points to the middle node when the fast pointer reaches the end of the list
+    return slow;
+}
+
+// Function to check if linked list is a palindrome or not
+bool isPalindrome(struct Node* head)
+{
+    // base cases
+    if (head == NULL) //No element, empty list
+        return false;
+    if (head->next == NULL) //Single element list
+        return true;
+        
+    // flag to indicate if number of nodes in the linked list is odd or not. It will be passed by reference to findMid()
+    bool odd = false;
+
+    // find second half of linked list
+    struct Node* mid = findMid(head, odd);
+
+    // if number of nodes is odd, advance mid as we'll start reversing from the node succeeding mid.
+    if (odd)
+        mid = mid->next;
+
+    // reverse the second half
+    reverse(&mid);
+
+    // compare first and second half
+    return compare(head, mid);
+}
+
+//T - TEST
+bool test1() //empty list
+{
+	char elements[] = {};
+    int n = sizeof(elements) / sizeof(elements[0]);
+    struct Node* head = NULL;
+    for (int i = 0; i < n; i++)
+        insert(&head, elements[i]);

```

Comment:
It looks like you copy and pasted your code infrastructure here even though it doesn't really apply since n is 0 and head = NULL.  I would simply omit all of this and call isPalindrome(NULL).

Reply:
For all of these functions, I would also get rid of the if statements and return what you expect.  If the list is a palindrome, simply 
`return isPalindrome(head)` otherwise `return !isPalindrom(head)`

Consider this any time you are going to do something like
if (boolCondition) return true